### PR TITLE
test: more gradient optimizer tests

### PR DESCRIPTION
**Pull request by novikov-alexander**

### CODE REVIEW

File: test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs [hunk 1]
```diff
@@ -67,6 +68,56 @@ public void TestBasic()
...
+        {
+            var dtype = GetTypeForNumericType<T>();
+
+            // train.GradientDescentOptimizer is V1 only API.
+            tf.Graph().as_default();
+            using (var sess = self.cached_session())
+            {
+                var var0 = tf.Variable(new[] { new[] { 1.0, 2.0 } }, dtype: dtype);
+                var var1 = tf.Variable(new[] { 3.0 }, dtype: dtype);
+                var x = tf.constant(
```
Comment (novikov-alexander) [inline test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs:81]: @Oceania2018 I can't initialize such a 2D constant. It either throws conversion exception or data allocation exception. Do I do anything wrong or there is a bug?

File: test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs [hunk 2]
```diff
@@ -115,5 +166,72 @@ public void TestTensorLearningRate()
...
+                var var0 = tf.Variable(new[] { 1.0, 2.0 }, dtype: dtype);
+                var var1 = tf.Variable(new[] { 3.0, 4.0 }, dtype: dtype);
+                var grads0 = tf.constant(new[] { 0.1, 0.1 }, dtype: dtype);
+                var grads1 = tf.constant(new[] { 0.01, 0.01 }, dtype: dtype);
+                var grads_and_vars = new[] {
+                    Tuple.Create(grads0, var0 as IVariableV1),
+                    Tuple.Create(grads1, var1 as IVariableV1)
+                };
+                var sgd_op = tf.train.GradientDescentOptimizer(3.0f)
+                    .apply_gradients(grads_and_vars, global_step: global_step);
```
Comment (novikov-alexander) [inline test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs:214]: @AsakusaRinne why does `apply_gradiens` take `System.Tuple` while `zip` produces `(T1, T2)`. Would it be better to replace or extend `apply_gradients` interface to support valuetuple as well?

File: test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs (hunk 1 above)
Comment (Wanglongzhi2001) [reply test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs:81]: Is it okay to float type? I'm not sure whether tf.net can initialize a 2D constant with double type, but it absolutly can when float type.

Comment (novikov-alexander) [reply test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs:81]: I just tried float type.
```
public static TensorProto make_tensor_proto(object values, TF_DataType dtype = TF_DataType.DtInvalid, Shape? shape = null, bool verify_shape = false, bool allow_broadcast = false)
{
   ...
   var len = dtype.get_datatype_size() * (int)shape.size;
   byte[] bytes = new byte[len];
   System.Buffer.BlockCopy(array, 0, bytes, 0, len);
```
`BlockCopy` expects that array contains only primitive values and throws there. 
Could you please give me an example where a float 2D constant works?

Comment (Wanglongzhi2001) [reply test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs:81]: Here is a example use float type:
https://github.com/SciSharp/TensorFlow.NET/blob/8cab730f98112a845461bb635f137e220f272adb/test/TensorFlowNET.Graph.UnitTest/ImageTest.cs#L54-L82

Comment (novikov-alexander) [reply test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs:81]: I see. It should be a 2D array `[,]` , not a jagged array `[][]`. Thanks.

File: test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs [hunk 3]
```diff
@@ -67,6 +68,51 @@ public void TestBasic()
...
+                sess.run(new[] { var0, var1 });
+                // Fetch params to validate initial values
+                self.assertAllCloseAccordingToType<T>(new[,] { { 1.0, 2.0 } }, self.evaluate<T[,]>(var0));
+                self.assertAllCloseAccordingToType(new[] { 3.0 }, self.evaluate<T[]>(var1));
+                // Run 1 step of sgd
+                sgd_op.run();
+                // Validate updated params
+                var np_pred = 1.0 * 4.0 + 2.0 * 5.0 + 3.0;
+                var np_grad = 2 * np_pred;
+                self.assertAllCloseAccordingToType(
```
Comment (novikov-alexander) [inline test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs:99]: @Wanglongzhi2001 Hm, now it calculates but the test doesn't pass. However, the code corresponds to TensorFlow original test. I have to check math there.

Comment (novikov-alexander) [reply test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs:99]: There just was a small typo, but I didn't have time to debug it :-D

**Comment by novikov-alexander** [comment]
Hi @AsakusaRinne  @Wanglongzhi2001 
I fixed the mistake. Please review when you have bandwidth
